vm/systemSmartContracts/governance.go

```diff
@@ -385,972 +394,573 @@ func (g *governanceContract) delegateVote(args *vmcommon.ContractCallInput) vmco
+		proposal.Passed = true
+		return nil
 	}
 
-	return proposal.EndVoteNonce
+	g.eei.AddReturnMessage("Proposal rejected")
+	proposal.Passed = false
+	return nil
 }
 
-// getGeneralProposal returns a proposal from storage
-func (g *governanceContract) getGeneralProposal(reference []byte) (*GeneralProposal, error) {
-	key := append([]byte(proposalPrefix), reference...)
-	marshaledData := g.eei.GetStorage(key)
-
+func (g *governanceContract) getActiveFundForDelegator(delegationAddress []byte, address []byte) (*big.Int, error) {
+	marshaledData := g.eei.GetStorageFromAddress(delegationAddress, address)
 	if len(marshaledData) == 0 {
-		return nil, vm.ErrProposalNotFound
+		return big.NewInt(0), nil
 	}
 
-	generalProposal := &GeneralProposal{}
-	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
+	dData := &DelegatorData{}
+	err := g.marshalizer.Unmarshal(dData, marshaledData)
 	if err != nil {
 		return nil, err
 	}
 
-	return generalProposal, nil
-}
-
-// proposalExists returns true if a proposal already exists
-func (g *governanceContract) proposalExists(reference []byte) bool {
-	key := append([]byte(proposalPrefix), reference...)
-	marshaledData := g.eei.GetStorage(key)
-	return len(marshaledData) > 0
-}
+	if len(dData.ActiveFund) == 0 {
+		return big.NewInt(0), nil
+	}
 
-// getValidProposal returns a proposal from storage if it exists, or it is still valid/in-progress
-func (g *governanceContract) getValidProposal(reference []byte) (*GeneralProposal, error) {
-	proposal, err := g.getGeneralProposal(reference)
+	marshaledData = g.eei.GetStorageFromAddress(delegationAddress, dData.ActiveFund)
+	activeFund := &Fund{Value: big.NewInt(0)}
+	err = g.marshalizer.Unmarshal(activeFund, marshaledData)
 	if err != nil {
 		return nil, err
 	}
 
-	currentNonce := g.eei.BlockChainHook().CurrentNonce()
-	if currentNonce < proposal.StartVoteNonce {
-		return nil, vm.ErrVotingNotStartedForProposal
-	}
-
-	if currentNonce > proposal.EndVoteNonce {
-		return nil, vm.ErrVotedForAnExpiredProposal
-	}
-
-	return proposal, nil
+	return activeFund.Value, nil
 }
 
-// isWhiteListed checks if an address is whitelisted
-func (g *governanceContract) isWhiteListed(address []byte) bool {
-	key := append([]byte(whiteListPrefix), address...)
-	marshaledData := g.eei.GetStorage(key)
+func (g *governanceContract) getTotalStake(validatorAddress []byte) (*big.Int, error) {
+	marshaledData := g.eei.GetStorageFromAddress(g.validatorSCAddress, validatorAddress)
 	if len(marshaledData) == 0 {
-		return false
+		return big.NewInt(0), nil
 	}
 
-	key = append([]byte(proposalPrefix), address...)
-	marshaledData = g.eei.GetStorage(key)
-	generalProposal := &GeneralProposal{}
-	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
+	validatorData := &ValidatorDataV2{}
+	err := g.marshalizer.Unmarshal(validatorData, marshaledData)
 	if err != nil {
-		return false
+		return nil, err
 	}
 
-	return generalProposal.Passed
+	return validatorData.TotalStakeValue, nil
```

Comment:
in some situations, this can return nil. Is this OK?

Reply:
there is no situation when that is nil. but added protection.